### PR TITLE
Implement `Keystore` spec for BLS12-381 secret keys

### PR DESCRIPTION
**Pull request by pawanjay176**
## Proposed Changes

Implement the `Keystore` standard [eip-2335](https://github.com/CarlBeek/EIPs/blob/bls_keystore/EIPS/eip-2335.md) for BLS12-381 secret keys.

Account paths [eip-2334](https://github.com/CarlBeek/EIPs/blob/bls_keystore/EIPS/eip-2335.md#path) is not implemented as we don't have/need `key-trees` currently. Can add it in a future PR.

### Todo

- [x] Decide on folder structure for storing keys.

### CODE REVIEW

**Comment by AgeManning** [comment]
I will likely get around to reviewing this tomorrow, or the weekend. In the meantime, @paulhauner  @michaelsproul what are your thoughts about the folder structure for storing keys. I personally don't have much preference.

**Comment by pawanjay176** [comment]
The way geth and parity do it currently is they have a `keystore` / `keys` folder which contains the keystore named as `UTC creation time + id` where `id` is a field in the keystore. 

We could have a similar setup along with some distinction between `validator` and `withdrawal` keystores but its not strictly required.

**Comment by AgeManning** [comment]
Yeah cool. They also tag the filename with the address (derived from the pubkey) and we could just use the publc key

**Comment by paulhauner** [comment]
As far as I can see, each validator will need:

1. One of these keystore files.
1. A message history database.
1. A lock file (maybe).

I think following the `UTC creation time + uuiid` convention for the keystore file is good. WRT to the folder, i'd be tempted to just use the `uuid`.

#### Example

```
.lighthouse/
  - validators/
    - <uuid>/
      - UTC-<creation time>-<uuid>
      - validator-history-<uuid>.sqlite
      - lockfile
```

File: account_manager/src/keystore/checksum.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+use crypto::digest::Digest;
+use crypto::sha2::Sha256;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+/// Checksum module for `Keystore`.
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct ChecksumModule {
+    pub function: String,
+    pub params: BTreeMap<(), ()>, // TODO: need a better way to encode empty json object
```
Comment (paulhauner) [inline account_manager/src/keystore/checksum.rs]: Perhaps [`serde_json::Value`](https://github.com/serde-rs/json#operating-on-untyped-json-values) would work?

Comment (pawanjay176) [reply account_manager/src/keystore/checksum.rs]: Yup that's perfect. Thanks! :)

**Comment by pawanjay176** [comment]
Folder structure used currently:
```
.lighthouse/
  - validators/
    - <uuid>/
      - "voting"-<UTC creation time>-<uuid>
      - "withdrawal"-<UTC creation time>-<uuid>
      - validator-history-<uuid>.sqlite
      - lockfile
```
Note: Voting and withdrawal keystores have the same `uuid` to ensure correspondence for later withdrawals.

**Comment by paulhauner** [comment]
Could you please merge in master when you get a chance? :)

**Comment by pawanjay176** [comment]
@paulhauner Merged master. 

I'm not sure yet how to handle the `Keystore` in `ValidatorDirectory`. 
One option is to add `voting_keystore` and `withdrawal_keystore` in the `ValidatorDirectory` struct as `Option<Keystore>` and convert the `Keystore` to a `Keypair` when required.

**Comment by AgeManning** [comment]
I'm working towards v0.2.0. I can take this over. Are we still blocked on deciding folder structure?
@paulhauner @pawanjay176

**Comment by pawanjay176** [comment]
Folder structure we have in the PR now is:
 ```
.lighthouse/
  - validators/
    - <uuid>/
      - "voting"-<UTC creation time>-<uuid>
      - "withdrawal"-<UTC creation time>-<uuid>
      - ...
 ```
The only difference in the folder structure compared to the testnet would be having the `uuid` as the folder name instead of the public key.

**Comment by pawanjay176** [comment]
@AgeManning created a new PR #777 to maintain a clean commit history. This one got messed up with my old eth1 branch.

**Pull request closed without merging**